packages/asset-swapper/src/utils/quote_reporter.ts

```diff
@@ -0,0 +1,171 @@
+import { BigNumber } from '@0x/utils';
+import * as _ from 'lodash';
+
+import { ERC20BridgeSource, SignedOrder } from '..';
+import { MarketOperation } from '../types';
+
+import { CollapsedFill, DexSample, NativeCollapsedFill } from './market_operation_utils/types';
+
+/**
+ * Differentiates different sources of native 0x order
+ */
+export enum NativeOrderOrigin {
+    RfqtOrigin = 'Rfqt',
+    OrderbookOrigin = 'Orderbook',
+}
+
+/**
+ * Represents a report of non-native liquidity source
+ */
+export interface BridgeReportSource {
+    liquiditySource: Exclude<ERC20BridgeSource, ERC20BridgeSource.Native>;
+    makerAmount: BigNumber;
+    takerAmount: BigNumber;
+}
+
+interface NativeLiquiditySource {
+    liquiditySource: ERC20BridgeSource.Native;
+    makerAmount: BigNumber;
+    takerAmount: BigNumber;
+    nativeOrder: SignedOrder;
+}
+
+export interface OrderbookReportSource extends NativeLiquiditySource {
+    nativeOrderOrigin: NativeOrderOrigin.OrderbookOrigin;
+}
+export interface RfqtReportSource extends NativeLiquiditySource {
+    nativeOrderOrigin: NativeOrderOrigin.RfqtOrigin;
+    makerUri: string;
+}
+/**
+ * Standardized representation of liquidity sources used in QuoteReport
+ */
+export type QuoteReportSource = BridgeReportSource | OrderbookReportSource | RfqtReportSource;
+
+/**
+ * Object representing a report of all liquidity sources
+ * considered during SwapQuote generation, as well as the path that
+ * was ultimately generated.
+ */
+export interface QuoteReport {
+    sourcesConsidered: QuoteReportSource[];
+    sourcesDelivered: QuoteReportSource[];
+}
+
+/**
+ * Class responsible for keeping track of all liquidity sources considered
+ * and the actual path generated while generating a SwapQuote
+ */
+export class QuoteReporter {
+    private readonly _marketOperation: MarketOperation;
+    private _orderbookReportSources: OrderbookReportSource[];
+    private _rfqtReportSources: RfqtReportSource[];
+    private _bridgeReportSources: BridgeReportSource[];
+    private _sourcesDelivered: QuoteReportSource[];
+
+    constructor(marketOperation: MarketOperation) {
+        this._orderbookReportSources = [];
+        this._rfqtReportSources = [];
+        this._bridgeReportSources = [];
+        this._sourcesDelivered = [];
+        this._marketOperation = marketOperation;
+    }
+
+    public trackDexSamples(dexSamples: DexSample[]): void {
+        this._bridgeReportSources = dexSamples.map(ds => this._dexSampleToBridgeReportSource(ds));
+    }
+
+    public trackOrderbookOrders(orderbookOrders: SignedOrder[]): void {
+        this._orderbookReportSources = orderbookOrders.map(oo => {
+            const orderbookReportSource: OrderbookReportSource = {
+                makerAmount: oo.makerAssetAmount,
+                takerAmount: oo.takerAssetAmount,
+                liquiditySource: ERC20BridgeSource.Native,
+                nativeOrderOrigin: NativeOrderOrigin.OrderbookOrigin,
```

Comment:
I think we want fillableAmounts here, i.e if the order is partially filled or the maker lacks a full balance

Reply:
2e2be4743937f56c2347f7845c028dc9ce4d7ac4